graspy/align/seedless_procrustes.py

```diff
@@ -0,0 +1,359 @@
+
+    def _optimal_transport(self, X, Y, Q):
+        # "E step" of the SeedlessProcrustes.
+        n, d = X.shape
+        m, _ = Y.shape
+        # initialize probability mass arrays & the cost matrix ; run sinkhorn
+        probability_mass_X = np.ones(n) / n
+        probability_mass_Y = np.ones(m) / m
+        cost_matrix = (
+            np.linalg.norm((X @ Q).reshape(n, 1, d) - Y.reshape(1, m, d), axis=2) ** 2
+        )
+        P = ot.sinkhorn(
+            a=probability_mass_X,
+            b=probability_mass_Y,
+            M=cost_matrix,
+            reg=self.optimal_transport_lambda,
+            numItermax=self.optimal_transport_eps,
+            stopThr=self.optimal_transport_eps,
+        )
+        return P
+
+    def _procrustes(self, X, Y, P):
+        # "M step" of the SeedlessProcurstes.
+        aligner = OrthogonalProcrustes()
+        Q = aligner.fit(X, P @ Y).Q_
+        return Q
+
+    def _iterative_ot(self, X, Y, Q):
+        # this P is not used. it is set to default in case numreps=0
+        P = np.ones((X.shape[0], Y.shape[0])) / (X.shape[0] * Y.shape[0])
+        for i in range(self.iterative_num_reps):
+            P = self._optimal_transport(X, Y, Q)
+            Q = self._procrustes(X, Y, P)
+        return P, Q
+
+    def _compute_objective(self, X, Y, Q=None, P=None):
+        if Q is None:
+            Q = self.Q_
+        if P is None:
+            P = self.P_
+        return np.linalg.norm(X @ Q - P @ Y, ord="fro")
+
+    def fit(self, X, Y):
+        """
+        Uses the two datasets to learn the matrix `self.Q_` that aligns the
+        first dataset with the second.
+
+        Parameters
+        ----------
+        X : np.ndarray, shape (n, d)
+            First dataset of vectors. These vectors need to have same number of
+            dimensions as ones in `Y`, but the number of vectors can differ.
+
+        Y : np.ndarray, shape (m, d)
+            Second dataset of vectors. These vectors need to have same number
+            of dimensions as ones in `X`, but the number of vectors can differ.
+
+        Returns
+        -------
+        self : returns an instance of self
+        """
+        X, Y = self._check_datasets(X, Y)
+        n, d = X.shape
+        m, _ = Y.shape
+
+        if self.init == "2d":
+            P_matrices = np.zeros((2 ** d, n, m))
+            Q_matrices = np.zeros((2 ** d, d, d))
+            objectives = np.zeros(2 ** d)
+            # try 2^d different initializations
+            for i in range(2 ** d):
+                initial_Q = _sign_flip_matrix_from_int(i, d)
+                P_matrices[i], Q_matrices[i] = P, Q = self._iterative_ot(
+                    X, Y, initial_Q
+                )
+                objectives[i] = self._compute_objective(X, Y, Q, P)
+            # pick the best one, using the objective function value
+            best = np.argmin(objectives)
+            self.initial_Q = _sign_flip_matrix_from_int(best, d)
```

Comment:
overloading this input param still gives me the heebie jeebies. 

i forget what we said in a previous thread, but would like to find a solution. 

one (possibly really bad) option is `self.initial_Q_` being separate from `self.initial_Q` 

another is we don't store anything about `initial_Q`s that were computed by us

another is we just need to come up with a damn name that doesn't confuse the two... i am having a surprisingly hard time doing so.

Reply:
also if we are in a rush to merge in can always kick this stone down the road. if we don't include this as an attribute and then find out we need it later... could put it in with some other name

Reply:
for example, many things in sklearn do random inits and then don't give you any info on what happened https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html

Reply:
so maybe we are overthinking this as researchy people

Reply:
so, an important reminder: it can only overload `None`'s, because it only happens when `initial_Q` is either adaptively selected (via 2^d / sign-flips), OR is inferred from the `initial_P`. by no means a panacea, but shouldn't lead to completely inexplicable behavior of "why did the matrix i pass mutate? o_O".

i've considered `initial_Q_`, may have even mentioned it above, but this does feel like a way to cause a ton of confusion.  parallel universe I, who uses graspologic, but did not contribute to, it would certainly think "who could have possibly think that this is a good idea to have `initial_Q` and `initial_Q_`?!?

i believe that i used to not have this behavior previously, but i needed to figure out the initial Q too many times, that i eventually just exposed it. maybe the need to know it only exists because we are still in the development stages of this methodology, and applications of this wouldn't care about it, not sure.

but for purposes like the triaging mentioned above, i certainly strongly feel that it should be exposed at least somehow, because unlike the Kmeans initis, at least i, happen to care about this sometimes, and if i didn't have access, i'd have to manually cycle all 2^d's and plug them in as customs. now, inside the .fit(), having initial_Q the parameter and initial_Q the attribute to be the same seems somewhat elegant, but i understand your concerns with sk-learn compliance. i certainly feel strongly that we should keep it as an attribute, but you can convince me that we should instead call it `initial_Q_used_`"(which does feels cringey to me though).

Reply:
`selected_initial_Q_` 

only stored if not "custom"

Reply:
sure. so do _not_ store it if it was set to identity if nothing was provided, or initial P was provided and initial Q was inferred?

Reply:
up to you

Reply:
can i just store it always then? if initial_Q is provided, it will be equal to a copy of that?

Reply:
i guess whatever you think makes the most sense, we can always modify later if need be.